File: tests/fixtures/real_world/javascript/taint/safe_ssrf_allowlist.js
var express = require('express');
var app = express();

var ALLOWED_HOSTS = ['api.example.com', 'cdn.example.com'];

app.get('/proxy', function(req, res) {
    var url = req.query.url;
    var parsed = new URL(url);
    if (ALLOWED_HOSTS.indexOf(parsed.hostname) === -1) {
        return res.status(403).send('Forbidden');
    }
    fetch('https://' + parsed.hostname + parsed.pathname).then(function(response) {
        return response.text();
    }).then(function(body) {
        res.send(body);
    });
});

File: tests/fixtures/real_world/python/taint/safe_ssrf_allowlist.py
from flask import Flask, request
import httpx

app = Flask(__name__)
ALLOWED_HOSTS = ['api.example.com', 'cdn.example.com']

@app.route('/proxy')
def proxy():
    url = request.args.get('url')
    from urllib.parse import urlparse
    host = urlparse(url).hostname
    if host not in ALLOWED_HOSTS:
        return 'Forbidden', 403
    resp = httpx.get(url)
    return resp.text
